chdb/dbapi/cursors.py

```diff
@@ -99,6 +99,42 @@ def _escape_args(self, args, conn):
             # Worst case it will throw a Value error
             return conn.escape(args)
 
+    def _format_query(self, query, args, conn):
+        """Format query with arguments supporting ? and % placeholders."""
+        if args is None:
+            return query
+
+        if isinstance(args, (tuple, list)) and '?' in query:
+            escaped_args = self._escape_args(args, conn)
+            result = []
+            arg_index = 0
+            in_string = False
+            quote_char = None
+
+            for i, char in enumerate(query):
+                # Track string literals to avoid replacing ? inside strings
+                if not in_string and char in ("'", '"'):
+                    in_string = True
+                    quote_char = char
+                elif in_string and char == quote_char:
+                    # Check if it's an escaped quote
+                    if i == 0 or query[i-1] != '\\':
+                        in_string = False
+                        quote_char = None
+
+                # Only replace ? outside of string literals
+                if char == '?' and not in_string and arg_index < len(escaped_args):
+                    result.append(str(escaped_args[arg_index]))
+                    arg_index += 1
+                else:
+                    result.append(char)
+
+            return ''.join(result)
+        elif '%' in query:
```

Comment:
It seems a bit unrigorous

Reply:
The `%s` placeholder retained the old logic (without considering escaping, in strings, etc.). Has been fixed to be the same as the handling for the `?`.